File: server/models/user_model.js
const mongoose = require('mongoose');
const Schema = mongoose.Schema;
const bcrypt = require('bcrypt');
//const shoppingHistory = require('./shopping_history_model');
const receiptSchema = require('./receipt_model');

var userSchema = new mongoose.Schema({

    email: {
        type: String,
        require: true
    },

    google_id: {
        type: String
    },

    first_name: {
        type: String,
        require: true
    },

    last_name: {
        type: String,
        require: true
    },

    phone_number: {
        type: String,
        
    },

    image: {
        type: String
    },

    password: {
        type: String,
        
    },

    shoppingHistory:
    //  shoppingHistoryItemsSchema
    {
        receipt: [{
            type: {
                receiptSchema
            }
            // date: {
            //     type: Date,
            //     required: true
            // },
            // netTotal: {
            //     type: Number,
            //     required: true
            // },
            // gst: {
            //     type: Number,
            //     required: true
            // },
            // isDeleted: {
            //     type: Boolean,
            //     defaults: false
            // },
            // items: [
            //     {
            //         type: shoppingHistoryItemsSchema
            //     }
            // ]
            // [{
            //     productID: {
            //         type: String,
            //         required: true
            //     },
            //     productQuantity: {
            //         type: Number,
            //         required: true
            //     },
            //     grossTotal: {
            //         type: Number,
            //         required: true
            //     },
            //     isDeleted: {
            //         type: Boolean,
            //         defaults: false
            //     }
            // }]
        }]
    }

});

userSchema.pre('save', async function (next) {
    
    try {

         if (this.isModified('password') || this.isNew) {
            const salt = await bcrypt.genSalt(10)
            const hashedPassword = await bcrypt.hash(this.password, salt)
            this.password = hashedPassword
            next()
         }
    }
    catch (error)
     {
        next(error)
    }

})

userSchema.methods.isValidPassword = async function (password) {
    try {
        return await bcrypt.compare(password, this.password)

    }
    catch (error) {
        throw error
    }
}

userSchema.post('save', async function (next) {
    try {
        console.log("called after saving a user")
    }
    catch (error) {
        next(error)
    }

})
const User = mongoose.model('user', userSchema);
module.exports = User; // this is done so you can use User inside any of your files

File: lib/models/user_model.dart
import 'package:troll_e/models/receipt_object_model.dart';
import 'package:troll_e/models/temp_receipt_model.dart';
import 'package:troll_e/models/shopping_history.dart';
import 'package:troll_e/models/shopping_history.dart';

class UserModel {
  String? userId;
  final String email;
  final String firstName;
  final String lastName;
  final String phoneNumber;
  String? googleId;
  String password;
  String? image;
  // var shoppingHistory;
  List<ReceiptObject>? shoppingHistory;

  UserModel({
    this.userId,
    required this.email,
    required this.firstName,
    required this.lastName,
    required this.password,
    required this.phoneNumber,
    this.image,
    // this.shoppingHistory,
     this.shoppingHistory,
  });

  static UserModel fromJson(Map<String, dynamic> json) {
    print("inside fromjson of usermodel");
    var shoppingHistoryJson = json['shoppingHistory']['receipt'] as List<dynamic>;
    var shoppingHistory = shoppingHistoryJson.map((receiptJson) => ReceiptObject.fromJson(receiptJson)).toList();
    return UserModel(
        userId: json['_id'],
        email: json['email'],
        firstName: json['first_name'],
        lastName: json['last_name'],
        password: json['password'],
        phoneNumber: json['phone_number'],
        image: json['image'] as String? ?? '',
        shoppingHistory: shoppingHistory);

  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = <String, dynamic>{};
    data['email'] = email;
    data['first_name'] = firstName;
    data['last_name'] = lastName;
    data['password'] = password;
    data['phone_number'] = phoneNumber;
    data['image']=image ;
    //data['shoppingHistory'] = {'receipt': this.shoppingHistory.map((v) => v.toJson()).toList(),
    //};
    return data;
  }
}
